cpp/src_prims/distance/euclidean.cuh

```diff
@@ -18,15 +18,168 @@
+ * @tparam AccT   accumulation data-type
+ * @tparam OutT   output data-type (for C and D matrices)
+ * @tparam IdxT   index data-type
+ * @tparam Veclen number of k-elements loaded by each thread for every LDG call
+ *                it makes. check contractions.cuh for details.
+ * @tparam FinalLambda the final lambda called on final distance value
+ * @param[in]     x input matrix
+ * @param[in]     y input matrix
+ * @param[in]     xn row norms of input matrix A.
+ * @param[in]     yn row norms of input matrix B.
+ * @param[in]     m number of rows of A and C/D
+ * @param[in]     n number of columns of B and C/D
+ * @param[in]     k number of cols of A and rows of B
+ * @param[in]     sqrt if the square root is computed or not
+ * @param[output] pD output matrix
+ * @param fin_op  the final gemm epilogue lambda
+*  @param stream  cuda stream to launch cuda operations.
+ */
+template <typename DataT, typename AccT, typename OutT, typename IdxT,
+          int VecLen, typename FinalLambda>
+void euclideanExpImpl(const DataT *x, const DataT *y, const DataT *xn,
+                      const DataT *yn, IdxT m, IdxT n, IdxT k, bool sqrt,
+                      OutT *dOutput, FinalLambda fin_op, cudaStream_t stream) {
+  typedef typename raft::linalg::Policy4x4<DataT, VecLen>::Policy Policy;
+  dim3 grid(raft::ceildiv<int>(m, Policy::Mblk),
+            raft::ceildiv<int>(n, Policy::Nblk));
+  dim3 blk(Policy::Nthreads);
+
+  // Accumulation operation lambda
+  auto core_lambda = [] __device__(AccT & acc, DataT & x, DataT & y) {
+    acc += x * y;
+  };
+
+  // epilogue operation lambda for final value calculation
+  auto epilog_lambda = [xn, yn, m, n, k] __device__(
+                         AccT acc[Policy::AccRowsPerTh][Policy::AccColsPerTh],
+                         DataT * sxNorm, DataT * syNorm) {
+    __syncthreads();  // so that we can safely reuse smem
+
+    // Load x & y norms required by this threadblock in shmem buffer
+    for (int i = threadIdx.x; i < Policy::Mblk; i += Policy::Nthreads) {
+      auto idx = blockIdx.x * Policy::Mblk + i;
+      sxNorm[i] = idx < m ? xn[idx] : 0;
+    }
+    for (int i = threadIdx.x; i < Policy::Nblk; i += Policy::Nthreads) {
+      auto idx = blockIdx.y * Policy::Nblk + i;
+      syNorm[i] = idx < n ? yn[idx] : 0;
+    }
+
+    __syncthreads();
+    DataT regxn[Policy::AccRowsPerTh], regyn[Policy::AccColsPerTh];
+#pragma unroll
+    for (int i = 0; i < Policy::AccRowsPerTh; ++i) {
+      regxn[i] =
+        sxNorm[i * Policy::AccThRows + (threadIdx.x / Policy::AccThCols)];
+    }
+#pragma unroll
+    for (int i = 0; i < Policy::AccColsPerTh; ++i) {
+      regyn[i] =
+        syNorm[i * Policy::AccThCols + (threadIdx.x % Policy::AccThCols)];
+    }
+#pragma unroll
+    for (int i = 0; i < Policy::AccRowsPerTh; ++i) {
+#pragma unroll
+      for (int j = 0; j < Policy::AccColsPerTh; ++j) {
+        acc[i][j] = regxn[i] + regyn[j] - (DataT)2.0 * acc[i][j];
+      }
+    }
+  };
```

Comment:
this seems to be a recurring pattern with cosine and euclidean distances. How about moving this as a method inside pairwise distance class?

Reply:
yes makes sense.